# **Proyecto 2. Análisis de la Ejecución de Ingresos Públicos en Brasil**

- Librerías y configuración

In [1]:
# Para tratamiento de datos
import numpy as np
import pandas as pd
# Para visualizaciones
import seaborn as sns 
import matplotlib.pyplot as plt
# Para generar todas las combinaciones posibles
import itertools
# Para guardar DataFrames en Excel
from pandas import ExcelWriter
# Para gestión de fechas
from datetime import datetime
# Ignorar warnings
import warnings
warnings.filterwarnings("ignore")
# Configuración para poder visualizar todas las columnas de los DataFrames
pd.set_option('display.max_columns', None) 
# importamos funciones de soporte
from src import soporte_limpieza as sl

## **Fase 1: Unión de Conjuntos de Datos**

- Carga de archivos

In [2]:
# cargamos los archivos
# tenemos que añadir el sep = ";" porque los archivos están separados por ; y no por ,
# añadimos el encoding = "latin-1" porque como es un archivo en portugués, pandas no lo lee bien
df_2013 = pd.read_csv("datos/datos-2013.csv", sep = ";", encoding = "latin-1")
df_2014 = pd.read_csv("datos/datos-2014.csv", sep = ";", encoding = "latin-1")
df_2015 = pd.read_csv("datos/datos-2015.csv", sep = ";", encoding = "latin-1")
df_2016 = pd.read_csv("datos/datos-2016.csv", sep = ";", encoding = "latin-1")
df_2017 = pd.read_csv("datos/datos-2017.csv", sep = ";", encoding = "latin-1")
df_2018 = pd.read_csv("datos/datos-2018.csv", sep = ";", encoding = "latin-1")
df_2019 = pd.read_csv("datos/datos-2019.csv", sep = ";", encoding = "latin-1")
df_2020 = pd.read_csv("datos/datos-2020.csv", sep = ";", encoding = "latin-1")
df_2021 = pd.read_csv("datos/datos-2021.csv", sep = ";", encoding = "latin-1")

- Exploramos cada archivo individualmente

In [3]:
df_2013.shape

(4498, 16)

In [4]:
reporte_2013 = sl.reporte(df_2013)
reporte_2013

index  valores_nulos  porcentaje_nulos tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0               0.0     int64
1         NOME ÓRGÃO SUPERIOR              0               0.0    object
2                CÓDIGO ÓRGÃO              0               0.0     int64
3                  NOME ÓRGÃO              0               0.0    object
4      CÓDIGO UNIDADE GESTORA              0               0.0     int64
5        NOME UNIDADE GESTORA              0               0.0    object
6         CATEGORIA ECONÔMICA              0               0.0    object
7              ORIGEM RECEITA              0               0.0    object
8             ESPÉCIE RECEITA              0               0.0    object
9                DETALHAMENTO              0               0.0    object
10  VALOR PREVISTO ATUALIZADO              0               0.0    object
11              VALOR LANÇADO              0               0.0    object
12            VALOR REALIZADO              0               0.0    object
13       PERCENTUAL REALIZADO              0               0.0    object
14            DATA LANÇAMENTO              0               0.0    object
15              ANO EXERCÍCIO              0               0.0     int64

No hay valores nulos y la mayoría de las columnas son de tipo object. Esto no acaba de encajar con el tipo de datos que deberían contener las columnas, es posible que las columnas de VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO' y 'PERCENTUAL REALIZADO' deban ser de tipo numérico y 'DATA LANÇAMENTO' de tipo fecha.  
Pasamos a hacer un describe de las columnas tipo object para tener un recuento de valores únicos

In [5]:
df_2013.describe(include ="O").T

count unique  \
NOME ÓRGÃO SUPERIOR        4498     25   
NOME ÓRGÃO                 4498    266   
NOME UNIDADE GESTORA       4498    279   
CATEGORIA ECONÔMICA        4498      5   
ORIGEM RECEITA             4498     15   
ESPÉCIE RECEITA            4498     44   
DETALHAMENTO               4498    704   
VALOR PREVISTO ATUALIZADO  4498   2788   
VALOR LANÇADO              4498     52   
VALOR REALIZADO            4498   3738   
PERCENTUAL REALIZADO       4498    578   
DATA LANÇAMENTO            4498      1   

                                                                         top  \
NOME ÓRGÃO SUPERIOR                                   Ministério da Educação   
NOME ÓRGÃO                 Ministério da Economia - Unidades com vínculo ...   
NOME UNIDADE GESTORA                 SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
CATEGORIA ECONÔMICA                                       Receitas Correntes   
ORIGEM RECEITA                                     Outras Receitas Correntes   
ESPÉCIE RECEITA                                          Receita de Serviços   
DETALHAMENTO                    RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES   
VALOR PREVISTO ATUALIZADO                                               0,00   
VALOR LANÇADO                                                           0,00   
VALOR REALIZADO                                                         0,00   
PERCENTUAL REALIZADO                                                    0,00   
DATA LANÇAMENTO                                                   31/12/2013   

                           freq  
NOME ÓRGÃO SUPERIOR        1833  
NOME ÓRGÃO                  494  
NOME UNIDADE GESTORA        415  
CATEGORIA ECONÔMICA        4001  
ORIGEM RECEITA             1982  
ESPÉCIE RECEITA             978  
DETALHAMENTO                214  
VALOR PREVISTO ATUALIZADO  1625  
VALOR LANÇADO              4447  
VALOR REALIZADO             704  
PERCENTUAL REALIZADO       2320  
DATA LANÇAMENTO            4498

In [6]:
df_2013.describe().T

count           mean            std       min  \
CÓDIGO ÓRGÃO SUPERIOR   4498.0   31729.657626   10691.972258   20000.0   
CÓDIGO ÓRGÃO            4498.0   31447.970431   10798.289531   20101.0   
CÓDIGO UNIDADE GESTORA  4498.0  233059.797910  141561.057724  110005.0   
ANO EXERCÍCIO           4498.0    2013.000000       0.000000    2013.0   

                             25%       50%       75%       max  
CÓDIGO ÓRGÃO SUPERIOR    25000.0   26000.0   36000.0   81000.0  
CÓDIGO ÓRGÃO             25916.0   26292.0   35000.0   91214.0  
CÓDIGO UNIDADE GESTORA  153173.0  160075.0  254420.0  913001.0  
ANO EXERCÍCIO             2013.0    2013.0    2013.0    2013.0

El describe() de las columnas numéricas en este caso no aporta nada, ya que los códigos son identificadores de cada organización y el año es el mismo en todos los registros.

In [7]:
df_2014.shape

(4553, 16)

In [8]:
reporte_2014 = sl.reporte(df_2014)
reporte_2014

index  valores_nulos  porcentaje_nulos tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0               0.0     int64
1         NOME ÓRGÃO SUPERIOR              0               0.0    object
2                CÓDIGO ÓRGÃO              0               0.0     int64
3                  NOME ÓRGÃO              0               0.0    object
4      CÓDIGO UNIDADE GESTORA              0               0.0     int64
5        NOME UNIDADE GESTORA              0               0.0    object
6         CATEGORIA ECONÔMICA              0               0.0    object
7              ORIGEM RECEITA              0               0.0    object
8             ESPÉCIE RECEITA              0               0.0    object
9                DETALHAMENTO              0               0.0    object
10  VALOR PREVISTO ATUALIZADO              0               0.0    object
11              VALOR LANÇADO              0               0.0    object
12            VALOR REALIZADO              0               0.0    object
13       PERCENTUAL REALIZADO              0               0.0    object
14            DATA LANÇAMENTO              0               0.0    object
15              ANO EXERCÍCIO              0               0.0     int64

In [9]:
df_2014.describe(include="O").T

count unique  \
NOME ÓRGÃO SUPERIOR        4553     25   
NOME ÓRGÃO                 4553    267   
NOME UNIDADE GESTORA       4553    278   
CATEGORIA ECONÔMICA        4553      5   
ORIGEM RECEITA             4553     15   
ESPÉCIE RECEITA            4553     46   
DETALHAMENTO               4553    708   
VALOR PREVISTO ATUALIZADO  4553   2740   
VALOR LANÇADO              4553     72   
VALOR REALIZADO            4553   3726   
PERCENTUAL REALIZADO       4553    546   
DATA LANÇAMENTO            4553      1   

                                                                         top  \
NOME ÓRGÃO SUPERIOR                                   Ministério da Educação   
NOME ÓRGÃO                 Ministério da Economia - Unidades com vínculo ...   
NOME UNIDADE GESTORA                 SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
CATEGORIA ECONÔMICA                                       Receitas Correntes   
ORIGEM RECEITA                                     Outras Receitas Correntes   
ESPÉCIE RECEITA                                          Receita de Serviços   
DETALHAMENTO                    RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES   
VALOR PREVISTO ATUALIZADO                                               0,00   
VALOR LANÇADO                                                           0,00   
VALOR REALIZADO                                                         0,00   
PERCENTUAL REALIZADO                                                    0,00   
DATA LANÇAMENTO                                                   31/12/2014   

                           freq  
NOME ÓRGÃO SUPERIOR        1841  
NOME ÓRGÃO                  482  
NOME UNIDADE GESTORA        382  
CATEGORIA ECONÔMICA        4095  
ORIGEM RECEITA             2050  
ESPÉCIE RECEITA             973  
DETALHAMENTO                216  
VALOR PREVISTO ATUALIZADO  1706  
VALOR LANÇADO              4482  
VALOR REALIZADO             757  
PERCENTUAL REALIZADO       2428  
DATA LANÇAMENTO            4553

In [10]:
df_2015.shape

(4523, 16)

In [11]:
reporte_2015 = sl.reporte(df_2015)
reporte_2015

index  valores_nulos  porcentaje_nulos tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0               0.0     int64
1         NOME ÓRGÃO SUPERIOR              0               0.0    object
2                CÓDIGO ÓRGÃO              0               0.0     int64
3                  NOME ÓRGÃO              0               0.0    object
4      CÓDIGO UNIDADE GESTORA              0               0.0     int64
5        NOME UNIDADE GESTORA              0               0.0    object
6         CATEGORIA ECONÔMICA              0               0.0    object
7              ORIGEM RECEITA              0               0.0    object
8             ESPÉCIE RECEITA              0               0.0    object
9                DETALHAMENTO              0               0.0    object
10  VALOR PREVISTO ATUALIZADO              0               0.0    object
11              VALOR LANÇADO              0               0.0    object
12            VALOR REALIZADO              0               0.0    object
13       PERCENTUAL REALIZADO              0               0.0    object
14            DATA LANÇAMENTO              0               0.0    object
15              ANO EXERCÍCIO              0               0.0     int64

In [12]:
df_2015.describe(include="O").T

count unique  \
NOME ÓRGÃO SUPERIOR        4523     25   
NOME ÓRGÃO                 4523    271   
NOME UNIDADE GESTORA       4523    288   
CATEGORIA ECONÔMICA        4523      4   
ORIGEM RECEITA             4523     14   
ESPÉCIE RECEITA            4523     41   
DETALHAMENTO               4523    707   
VALOR PREVISTO ATUALIZADO  4523   2648   
VALOR LANÇADO              4523     83   
VALOR REALIZADO            4523   3744   
PERCENTUAL REALIZADO       4523    457   
DATA LANÇAMENTO            4523      1   

                                                                         top  \
NOME ÓRGÃO SUPERIOR                                   Ministério da Educação   
NOME ÓRGÃO                 Ministério da Economia - Unidades com vínculo ...   
NOME UNIDADE GESTORA                 SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
CATEGORIA ECONÔMICA                                       Receitas Correntes   
ORIGEM RECEITA                                     Outras Receitas Correntes   
ESPÉCIE RECEITA                Multas administrativas, contratuais e judicia   
DETALHAMENTO                    RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES   
VALOR PREVISTO ATUALIZADO                                               0,00   
VALOR LANÇADO                                                           0,00   
VALOR REALIZADO                                                         0,00   
PERCENTUAL REALIZADO                                                    0,00   
DATA LANÇAMENTO                                                   31/12/2015   

                           freq  
NOME ÓRGÃO SUPERIOR        1803  
NOME ÓRGÃO                  481  
NOME UNIDADE GESTORA        385  
CATEGORIA ECONÔMICA        4088  
ORIGEM RECEITA             2076  
ESPÉCIE RECEITA             962  
DETALHAMENTO                225  
VALOR PREVISTO ATUALIZADO  1785  
VALOR LANÇADO              4441  
VALOR REALIZADO             711  
PERCENTUAL REALIZADO       2455  
DATA LANÇAMENTO            4523

In [13]:
df_2016.shape

(194533, 16)

In [14]:
reporte_2016 =sl.reporte(df_2016)
reporte_2016

index  valores_nulos  porcentaje_nulos tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0              0.00     int64
1         NOME ÓRGÃO SUPERIOR              0              0.00    object
2                CÓDIGO ÓRGÃO              0              0.00     int64
3                  NOME ÓRGÃO              0              0.00    object
4      CÓDIGO UNIDADE GESTORA              0              0.00     int64
5        NOME UNIDADE GESTORA              0              0.00    object
6         CATEGORIA ECONÔMICA              0              0.00    object
7              ORIGEM RECEITA              0              0.00    object
8             ESPÉCIE RECEITA              0              0.00    object
9                DETALHAMENTO              0              0.00    object
10  VALOR PREVISTO ATUALIZADO              0              0.00    object
11              VALOR LANÇADO              0              0.00    object
12            VALOR REALIZADO              0              0.00    object
13       PERCENTUAL REALIZADO              0              0.00    object
14            DATA LANÇAMENTO             82              0.04    object
15              ANO EXERCÍCIO              0              0.00     int64

In [15]:
df_2016.describe(include="O").T

count  unique  \
NOME ÓRGÃO SUPERIOR        194533      25   
NOME ÓRGÃO                 194533     269   
NOME UNIDADE GESTORA       194533     293   
CATEGORIA ECONÔMICA        194533       4   
ORIGEM RECEITA             194533      14   
ESPÉCIE RECEITA            194533      44   
DETALHAMENTO               194533     525   
VALOR PREVISTO ATUALIZADO  194533    1835   
VALOR LANÇADO              194533    1003   
VALOR REALIZADO            194533  129122   
PERCENTUAL REALIZADO       194533      29   
DATA LANÇAMENTO            194451     360   

                                                                         top  \
NOME ÓRGÃO SUPERIOR                                   Ministério da Educação   
NOME ÓRGÃO                 Ministério da Economia - Unidades com vínculo ...   
NOME UNIDADE GESTORA                 SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
CATEGORIA ECONÔMICA                                       Receitas Correntes   
ORIGEM RECEITA                                     Outras Receitas Correntes   
ESPÉCIE RECEITA                 Serviços Administrativos e Comerciais Gerais   
DETALHAMENTO                     SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   
VALOR PREVISTO ATUALIZADO                                               0,00   
VALOR LANÇADO                                                           0,00   
VALOR REALIZADO                                                         0,00   
PERCENTUAL REALIZADO                                                    0,00   
DATA LANÇAMENTO                                                   21/01/2016   

                             freq  
NOME ÓRGÃO SUPERIOR         71840  
NOME ÓRGÃO                  13749  
NOME UNIDADE GESTORA        11126  
CATEGORIA ECONÔMICA        186687  
ORIGEM RECEITA              70647  
ESPÉCIE RECEITA             54941  
DETALHAMENTO                30237  
VALOR PREVISTO ATUALIZADO  192639  
VALOR LANÇADO              193450  
VALOR REALIZADO              2972  
PERCENTUAL REALIZADO       194412  
DATA LANÇAMENTO              2095

In [16]:
df_2017.shape

(190479, 16)

In [17]:
reporte_2017 =sl.reporte(df_2017)
reporte_2017

index  valores_nulos  porcentaje_nulos tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0              0.00     int64
1         NOME ÓRGÃO SUPERIOR              0              0.00    object
2                CÓDIGO ÓRGÃO              0              0.00     int64
3                  NOME ÓRGÃO              0              0.00    object
4      CÓDIGO UNIDADE GESTORA              0              0.00     int64
5        NOME UNIDADE GESTORA              0              0.00    object
6         CATEGORIA ECONÔMICA              0              0.00    object
7              ORIGEM RECEITA              0              0.00    object
8             ESPÉCIE RECEITA              0              0.00    object
9                DETALHAMENTO              0              0.00    object
10  VALOR PREVISTO ATUALIZADO              0              0.00    object
11              VALOR LANÇADO              0              0.00    object
12            VALOR REALIZADO              0              0.00    object
13       PERCENTUAL REALIZADO              0              0.00    object
14            DATA LANÇAMENTO             86              0.05    object
15              ANO EXERCÍCIO              0              0.00     int64

In [18]:
df_2017.describe(include="O").T

count  unique  \
NOME ÓRGÃO SUPERIOR        190479      24   
NOME ÓRGÃO                 190479     268   
NOME UNIDADE GESTORA       190479     318   
CATEGORIA ECONÔMICA        190479       5   
ORIGEM RECEITA             190479      15   
ESPÉCIE RECEITA            190479      47   
DETALHAMENTO               190479     516   
VALOR PREVISTO ATUALIZADO  190479    1827   
VALOR LANÇADO              190479    1043   
VALOR REALIZADO            190479  128532   
PERCENTUAL REALIZADO       190479      24   
DATA LANÇAMENTO            190393     365   

                                                                         top  \
NOME ÓRGÃO SUPERIOR                                   Ministério da Educação   
NOME ÓRGÃO                 Ministério da Economia - Unidades com vínculo ...   
NOME UNIDADE GESTORA                 SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
CATEGORIA ECONÔMICA                                       Receitas Correntes   
ORIGEM RECEITA                                     Outras Receitas Correntes   
ESPÉCIE RECEITA                 Serviços Administrativos e Comerciais Gerais   
DETALHAMENTO                     SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   
VALOR PREVISTO ATUALIZADO                                               0,00   
VALOR LANÇADO                                                           0,00   
VALOR REALIZADO                                                         0,00   
PERCENTUAL REALIZADO                                                    0,00   
DATA LANÇAMENTO                                                   19/01/2017   

                             freq  
NOME ÓRGÃO SUPERIOR         70837  
NOME ÓRGÃO                  14158  
NOME UNIDADE GESTORA        11227  
CATEGORIA ECONÔMICA        182668  
ORIGEM RECEITA              68818  
ESPÉCIE RECEITA             51633  
DETALHAMENTO                29841  
VALOR PREVISTO ATUALIZADO  188589  
VALOR LANÇADO              189434  
VALOR REALIZADO              2800  
PERCENTUAL REALIZADO       190295  
DATA LANÇAMENTO              2111

In [19]:
df_2018.shape

(173944, 16)

In [20]:
reporte_2018 =sl.reporte(df_2018)
reporte_2018

index  valores_nulos  porcentaje_nulos tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0              0.00     int64
1         NOME ÓRGÃO SUPERIOR              0              0.00    object
2                CÓDIGO ÓRGÃO              0              0.00     int64
3                  NOME ÓRGÃO              0              0.00    object
4      CÓDIGO UNIDADE GESTORA              0              0.00     int64
5        NOME UNIDADE GESTORA              0              0.00    object
6         CATEGORIA ECONÔMICA              0              0.00    object
7              ORIGEM RECEITA              0              0.00    object
8             ESPÉCIE RECEITA              0              0.00    object
9                DETALHAMENTO              0              0.00    object
10  VALOR PREVISTO ATUALIZADO              0              0.00    object
11              VALOR LANÇADO              0              0.00    object
12            VALOR REALIZADO              0              0.00    object
13       PERCENTUAL REALIZADO              0              0.00    object
14            DATA LANÇAMENTO            115              0.07    object
15              ANO EXERCÍCIO              0              0.00     int64

In [21]:
df_2018.describe(include="O").T

count  unique  \
NOME ÓRGÃO SUPERIOR        173944      22   
NOME ÓRGÃO                 173944     260   
NOME UNIDADE GESTORA       173944     312   
CATEGORIA ECONÔMICA        173944       5   
ORIGEM RECEITA             173944      15   
ESPÉCIE RECEITA            173944      47   
DETALHAMENTO               173944     608   
VALOR PREVISTO ATUALIZADO  173944    1846   
VALOR LANÇADO              173944     828   
VALOR REALIZADO            173944  121803   
PERCENTUAL REALIZADO       173944      12   
DATA LANÇAMENTO            173829     365   

                                                                         top  \
NOME ÓRGÃO SUPERIOR                                   Ministério da Educação   
NOME ÓRGÃO                 Ministério da Economia - Unidades com vínculo ...   
NOME UNIDADE GESTORA                 SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
CATEGORIA ECONÔMICA                                       Receitas Correntes   
ORIGEM RECEITA                                           Receita de Serviços   
ESPÉCIE RECEITA                 Serviços Administrativos e Comerciais Gerais   
DETALHAMENTO                     SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   
VALOR PREVISTO ATUALIZADO                                               0,00   
VALOR LANÇADO                                                           0,00   
VALOR REALIZADO                                                         0,00   
PERCENTUAL REALIZADO                                                    0,00   
DATA LANÇAMENTO                                                   12/01/2018   

                             freq  
NOME ÓRGÃO SUPERIOR         65876  
NOME ÓRGÃO                  22844  
NOME UNIDADE GESTORA        21206  
CATEGORIA ECONÔMICA        165585  
ORIGEM RECEITA              58581  
ESPÉCIE RECEITA             51444  
DETALHAMENTO                29257  
VALOR PREVISTO ATUALIZADO  172042  
VALOR LANÇADO              173114  
VALOR REALIZADO              2715  
PERCENTUAL REALIZADO       173831  
DATA LANÇAMENTO              2032

In [22]:
df_2019.shape

(176828, 16)

In [23]:
reporte_2019 =sl.reporte(df_2019)
reporte_2019

index  valores_nulos  porcentaje_nulos tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0              0.00     int64
1         NOME ÓRGÃO SUPERIOR              0              0.00    object
2                CÓDIGO ÓRGÃO              0              0.00     int64
3                  NOME ÓRGÃO              0              0.00    object
4      CÓDIGO UNIDADE GESTORA              0              0.00     int64
5        NOME UNIDADE GESTORA              0              0.00    object
6         CATEGORIA ECONÔMICA              0              0.00    object
7              ORIGEM RECEITA              0              0.00    object
8             ESPÉCIE RECEITA              0              0.00    object
9                DETALHAMENTO              0              0.00    object
10  VALOR PREVISTO ATUALIZADO              0              0.00    object
11              VALOR LANÇADO              0              0.00    object
12            VALOR REALIZADO              0              0.00    object
13       PERCENTUAL REALIZADO              0              0.00    object
14            DATA LANÇAMENTO             86              0.05    object
15              ANO EXERCÍCIO              0              0.00     int64

In [24]:
df_2019.describe(include="O").T

count  unique  \
NOME ÓRGÃO SUPERIOR        176828      20   
NOME ÓRGÃO                 176828     249   
NOME UNIDADE GESTORA       176828     269   
CATEGORIA ECONÔMICA        176828       5   
ORIGEM RECEITA             176828      15   
ESPÉCIE RECEITA            176828      46   
DETALHAMENTO               176828     636   
VALOR PREVISTO ATUALIZADO  176828    1801   
VALOR LANÇADO              176828    1105   
VALOR REALIZADO            176828  124002   
PERCENTUAL REALIZADO       176828      48   
DATA LANÇAMENTO            176742     365   

                                                                         top  \
NOME ÓRGÃO SUPERIOR                                   Ministério da Educação   
NOME ÓRGÃO                 Ministério da Economia - Unidades com vínculo ...   
NOME UNIDADE GESTORA                 SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
CATEGORIA ECONÔMICA                                       Receitas Correntes   
ORIGEM RECEITA                                           Receita de Serviços   
ESPÉCIE RECEITA                 Serviços Administrativos e Comerciais Gerais   
DETALHAMENTO                     SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   
VALOR PREVISTO ATUALIZADO                                               0,00   
VALOR LANÇADO                                                           0,00   
VALOR REALIZADO                                                         0,00   
PERCENTUAL REALIZADO                                                    0,00   
DATA LANÇAMENTO                                                   31/01/2019   

                             freq  
NOME ÓRGÃO SUPERIOR         67104  
NOME ÓRGÃO                  23514  
NOME UNIDADE GESTORA        22389  
CATEGORIA ECONÔMICA        169210  
ORIGEM RECEITA              59957  
ESPÉCIE RECEITA             52747  
DETALHAMENTO                29143  
VALOR PREVISTO ATUALIZADO  174975  
VALOR LANÇADO              175649  
VALOR REALIZADO              3107  
PERCENTUAL REALIZADO       176448  
DATA LANÇAMENTO              2212

In [25]:
df_2020.shape

(142348, 16)

In [26]:
reporte_2020 = sl.reporte(df_2020)
reporte_2020

index  valores_nulos  porcentaje_nulos tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0              0.00     int64
1         NOME ÓRGÃO SUPERIOR              0              0.00    object
2                CÓDIGO ÓRGÃO              0              0.00     int64
3                  NOME ÓRGÃO              0              0.00    object
4      CÓDIGO UNIDADE GESTORA              0              0.00     int64
5        NOME UNIDADE GESTORA              0              0.00    object
6         CATEGORIA ECONÔMICA              0              0.00    object
7              ORIGEM RECEITA              0              0.00    object
8             ESPÉCIE RECEITA              0              0.00    object
9                DETALHAMENTO              0              0.00    object
10  VALOR PREVISTO ATUALIZADO              0              0.00    object
11              VALOR LANÇADO              0              0.00    object
12            VALOR REALIZADO              0              0.00    object
13       PERCENTUAL REALIZADO              0              0.00    object
14            DATA LANÇAMENTO             86              0.06    object
15              ANO EXERCÍCIO              0              0.00     int64

In [27]:
df_2020.describe(include="O").T

count  unique  \
NOME ÓRGÃO SUPERIOR        142348      20   
NOME ÓRGÃO                 142348     254   
NOME UNIDADE GESTORA       142348     270   
CATEGORIA ECONÔMICA        142348       5   
ORIGEM RECEITA             142348      15   
ESPÉCIE RECEITA            142348      45   
DETALHAMENTO               142348     625   
VALOR PREVISTO ATUALIZADO  142348    1951   
VALOR LANÇADO              142348    1579   
VALOR REALIZADO            142348  106992   
PERCENTUAL REALIZADO       142348      27   
DATA LANÇAMENTO            142262     362   

                                                                         top  \
NOME ÓRGÃO SUPERIOR                                   Ministério da Educação   
NOME ÓRGÃO                 Ministério da Economia - Unidades com vínculo ...   
NOME UNIDADE GESTORA                 SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
CATEGORIA ECONÔMICA                                       Receitas Correntes   
ORIGEM RECEITA                                     Outras Receitas Correntes   
ESPÉCIE RECEITA                 Serviços Administrativos e Comerciais Gerais   
DETALHAMENTO                     SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   
VALOR PREVISTO ATUALIZADO                                               0,00   
VALOR LANÇADO                                                           0,00   
VALOR REALIZADO                                                         0,00   
PERCENTUAL REALIZADO                                                    0,00   
DATA LANÇAMENTO                                                   20/01/2020   

                             freq  
NOME ÓRGÃO SUPERIOR         40888  
NOME ÓRGÃO                  22686  
NOME UNIDADE GESTORA        21074  
CATEGORIA ECONÔMICA        135172  
ORIGEM RECEITA              42127  
ESPÉCIE RECEITA             33271  
DETALHAMENTO                21029  
VALOR PREVISTO ATUALIZADO  140352  
VALOR LANÇADO              140661  
VALOR REALIZADO              3837  
PERCENTUAL REALIZADO       142103  
DATA LANÇAMENTO              2179

In [28]:
df_2021.shape

(134593, 16)

In [29]:
reporte_2021 = sl.reporte(df_2021)
reporte_2021

index  valores_nulos  porcentaje_nulos tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0              0.00     int64
1         NOME ÓRGÃO SUPERIOR              0              0.00    object
2                CÓDIGO ÓRGÃO              0              0.00     int64
3                  NOME ÓRGÃO              0              0.00    object
4      CÓDIGO UNIDADE GESTORA              0              0.00     int64
5        NOME UNIDADE GESTORA              0              0.00    object
6         CATEGORIA ECONÔMICA              0              0.00    object
7              ORIGEM RECEITA              0              0.00    object
8             ESPÉCIE RECEITA              0              0.00    object
9                DETALHAMENTO              0              0.00    object
10  VALOR PREVISTO ATUALIZADO              0              0.00    object
11              VALOR LANÇADO              0              0.00    object
12            VALOR REALIZADO              0              0.00    object
13       PERCENTUAL REALIZADO              0              0.00    object
14            DATA LANÇAMENTO            123              0.09    object
15              ANO EXERCÍCIO              0              0.00     int64

In [30]:
df_2021.describe(include="O").T

count unique  \
NOME ÓRGÃO SUPERIOR        134593     20   
NOME ÓRGÃO                 134593    256   
NOME UNIDADE GESTORA       134593    268   
CATEGORIA ECONÔMICA        134593      5   
ORIGEM RECEITA             134593     15   
ESPÉCIE RECEITA            134593     42   
DETALHAMENTO               134593    600   
VALOR PREVISTO ATUALIZADO  134593   1803   
VALOR LANÇADO              134593   1191   
VALOR REALIZADO            134593  99376   
PERCENTUAL REALIZADO       134593     17   
DATA LANÇAMENTO            134470    337   

                                                                         top  \
NOME ÓRGÃO SUPERIOR                                   Ministério da Educação   
NOME ÓRGÃO                 Ministério da Economia - Unidades com vínculo ...   
NOME UNIDADE GESTORA                 SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
CATEGORIA ECONÔMICA                                       Receitas Correntes   
ORIGEM RECEITA                                           Receita de Serviços   
ESPÉCIE RECEITA                 Serviços Administrativos e Comerciais Gerais   
DETALHAMENTO                     SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   
VALOR PREVISTO ATUALIZADO                                               0,00   
VALOR LANÇADO                                                           0,00   
VALOR REALIZADO                                                         0,00   
PERCENTUAL REALIZADO                                                    0,00   
DATA LANÇAMENTO                                                   23/04/2021   

                             freq  
NOME ÓRGÃO SUPERIOR         39869  
NOME ÓRGÃO                  21403  
NOME UNIDADE GESTORA        20277  
CATEGORIA ECONÔMICA        128090  
ORIGEM RECEITA              39917  
ESPÉCIE RECEITA             34121  
DETALHAMENTO                19532  
VALOR PREVISTO ATUALIZADO  132746  
VALOR LANÇADO              133403  
VALOR REALIZADO              3404  
PERCENTUAL REALIZADO       134516  
DATA LANÇAMENTO              1957

Los 3 primeros años (2013, 2014 y 2015) no tienen valores nulos, sin embargo, del 2016 en adelante encontramos valores nulos en la columna "DATA LANÇAMENTO", son pocos, entre el 0.04 y el 0.09 %.
Mientras que en los 3 primeros años esta fecha era la misma en todos los registros, a partir de 2016 hay diferentes valores de fecha por lo que se plantea el problema de cómo rellenar esos valores en el caso de que sea necesario. 

- Comprobamos que los tipos y nombres de las columnas son consistentes en todos los archivos
    

Para eso comparamos las columnas "index" y "tipo_dato" de los reportes que hemos generado. 

In [31]:
# Usamos una función que compara unas columnas de una lista de reportes para comprobar si son iguales. Devuelve true si son iguales y false si encuentra alguna diferencia
reportes = [reporte_2013, reporte_2014, reporte_2015, reporte_2016, reporte_2017, reporte_2018, reporte_2019, reporte_2020, reporte_2021]
columnas = ["index", "tipo_dato"]
sl.comparar_reportes(reportes, columnas)

True

- Unión de los archivos

La función de comparació nos ha dado como resultado True, lo que indicaría que en todos los dataframe las columnas se llaman igual y tienen el mismo tipo de datos, por lo que podemos unir todos los dataframe usando concat.

In [32]:
# Unimos todos los archivos usando concat y axis = 0 para que los una de forma vertical
datos_brasil = pd.concat([df_2013, df_2014, df_2015, df_2016, df_2017, df_2018, df_2019, df_2020, df_2021], axis =0)

## **Fase 2: Limpieza de Datos**

- Corrección de formatos

Debemos cambiar el tipo de dato de las columnas 'VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO' y 'PERCENTUAL REALIZADO' de object a float.  
Para ello, debemos empezar cambiando comas por puntos para que pueda leerse como número. Usamos una función que aplica str.replace a una lista de columnas

In [33]:
# Cambiamos comas por puntos en las columnas 'VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO' y 'PERCENTUAL REALIZADO'.
sl.reemplazar(datos_brasil, ['VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO','PERCENTUAL REALIZADO'], ",", ".")

CÓDIGO ÓRGÃO SUPERIOR       NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
0                       63000  Advocacia-Geral da União         63000   
1                       63000  Advocacia-Geral da União         63000   
2                       63000  Advocacia-Geral da União         63000   
3                       63000  Advocacia-Geral da União         63000   
4                       63000  Advocacia-Geral da União         63000   
...                       ...                       ...           ...   
134588                  20000  Presidência da República         24208   
134589                  20000  Presidência da República         24208   
134590                  20000  Presidência da República         24208   
134591                  20000  Presidência da República         24208   
134592                  20000  Presidência da República         24208   

                                               NOME ÓRGÃO  \
0       Advocacia-Geral da União - Unidades com víncul...   
1       Advocacia-Geral da União - Unidades com víncul...   
2       Advocacia-Geral da União - Unidades com víncul...   
3       Advocacia-Geral da União - Unidades com víncul...   
4       Advocacia-Geral da União - Unidades com víncul...   
...                                                   ...   
134588     Instituto Nacional de Tecnologia da Informação   
134589     Instituto Nacional de Tecnologia da Informação   
134590     Instituto Nacional de Tecnologia da Informação   
134591     Instituto Nacional de Tecnologia da Informação   
134592     Instituto Nacional de Tecnologia da Informação   

        CÓDIGO UNIDADE GESTORA                           NOME UNIDADE GESTORA  \
0                       110060  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
1                       110060  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
2                       110060  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
3                       110060  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
4                       110060  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   
...                        ...                                            ...   
134588                  243001  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   
134589                  243001  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   
134590                  243001  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   
134591                  243001  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   
134592                  243001  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   

        CATEGORIA ECONÔMICA             ORIGEM RECEITA  \
0        Receitas Correntes  Outras Receitas Correntes   
1        Receitas Correntes  Outras Receitas Correntes   
2        Receitas Correntes  Outras Receitas Correntes   
3        Receitas Correntes  Outras Receitas Correntes   
4        Receitas Correntes  Outras Receitas Correntes   
...                     ...                        ...   
134588  Receitas de Capital       Operações de Crédito   
134589   Receitas Correntes        Receita de Serviços   
134590   Receitas Correntes        Receita de Serviços   
134591   Receitas Correntes        Receita de Serviços   
134592   Receitas Correntes        Receita de Serviços   

                                      ESPÉCIE RECEITA  \
0       Bens, Direitos e Valores Incorporados ao Patr   
1         Indenizações, restituições e ressarcimentos   
2       Multas administrativas, contratuais e judicia   
3       Bens, Direitos e Valores Incorporados ao Patr   
4         Indenizações, restituições e ressarcimentos   
...                                               ...   
134588         Operações de crédito - mercado interno   
134589   Serviços Administrativos e Comerciais Gerais   
134590   Serviços Administrativos e Comerciais Gerais   
134591   Serviços Administrativos e Comerciais Gerais   
134592   Serviços Administrativos e Comerciais Gerais   

                                         DE

In [38]:
# normalizamos valores en la columna "revenue" 
# Primero reemplazamos los ceros por NaN
datos_brasil[['VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO','PERCENTUAL REALIZADO']] = datos_brasil[['VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO','PERCENTUAL REALIZADO']].replace(0,np.nan)
# A continuación convertimos lo que quede a número, forzando a que en caso de error lo convierta en NaN
datos_brasil = sl.a_numero(datos_brasil, ['VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO','PERCENTUAL REALIZADO'])
datos_brasil[['VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO','PERCENTUAL REALIZADO']]  = datos_brasil[['VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO','PERCENTUAL REALIZADO']] .fillna(0).astype(float)

In [39]:
sl.reporte(datos_brasil)

index  valores_nulos  porcentaje_nulos tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0              0.00     int64
1         NOME ÓRGÃO SUPERIOR              0              0.00    object
2                CÓDIGO ÓRGÃO              0              0.00     int64
3                  NOME ÓRGÃO              0              0.00    object
4      CÓDIGO UNIDADE GESTORA              0              0.00     int64
5        NOME UNIDADE GESTORA              0              0.00    object
6         CATEGORIA ECONÔMICA              0              0.00    object
7              ORIGEM RECEITA              0              0.00    object
8             ESPÉCIE RECEITA              0              0.00    object
9                DETALHAMENTO              0              0.00    object
10  VALOR PREVISTO ATUALIZADO              0              0.00   float64
11              VALOR LANÇADO              0              0.00   float64
12            VALOR REALIZADO              0              0.00   float64
13       PERCENTUAL REALIZADO              0              0.00   float64
14            DATA LANÇAMENTO            578              0.06    object
15              ANO EXERCÍCIO              0              0.00     int64

In [36]:
# Cambiamos el tipo de dato de las columnas 'VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO' y 'PERCENTUAL REALIZADO' a float usando .astype
#datos_brasil[['VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO','PERCENTUAL REALIZADO']] = datos_brasil[['VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO', 'VALOR REALIZADO','PERCENTUAL REALIZADO']].astype(float)
# Cambiamos el tipo de la columna DATA LANÇAMENTO a datetime
#datos_brasil["DATA LANÇAMENTO"] = pd.to_datetime(datos_brasil["DATA LANÇAMENTO"])
# Por último, como los códigos de las entidades no se usan en ninguna operación, los pasamos a string para que no generen ruido en los análisis estadísticos
#datos_brasil[["CÓDIGO ÓRGÃO SUPERIOR", "CÓDIGO ÓRGÃO", "CÓDIGO UNIDADE GESTORA"]] = datos_brasil[["CÓDIGO ÓRGÃO SUPERIOR", "CÓDIGO ÓRGÃO", "CÓDIGO UNIDADE GESTORA"]].astype(str)


In [37]:
# Vamos a eliminar espacios a principio y fin de las columnas tipo object y además, pondremos en formato título (primera letra de cada palabra en mayúscula) todas las de NOME...
# Eliminamos espacios
datos_brasil = sl.espacios(datos_brasil,['CÓDIGO ÓRGÃO SUPERIOR', 'NOME ÓRGÃO SUPERIOR', 'CÓDIGO ÓRGÃO',
       'NOME ÓRGÃO', 'CÓDIGO UNIDADE GESTORA', 'NOME UNIDADE GESTORA',
       'CATEGORIA ECONÔMICA', 'ORIGEM RECEITA', 'ESPÉCIE RECEITA',
       'DETALHAMENTO']) 
# Pasamos a formato título
datos_brasil = sl.titulo(datos_brasil,['NOME ÓRGÃO SUPERIOR', 'NOME ÓRGÃO', 'NOME UNIDADE GESTORA'])
datos_brasil.sample(3)

AttributeError: Can only use .str accessor with string values!

In [68]:
# Guardamos en formato pickle para mantener los tipos de dato
datos_brasil.to_pickle("datos/datos_brasil.pkl")

- Tratamiento de Valores Nulos

Comprobamos si hay valores duplicados

In [45]:
duplicados = datos_brasil[datos_brasil.duplicated()]
duplicados

Empty DataFrame
Columns: [CÓDIGO ÓRGÃO SUPERIOR, NOME ÓRGÃO SUPERIOR, CÓDIGO ÓRGÃO, NOME ÓRGÃO, CÓDIGO UNIDADE GESTORA, NOME UNIDADE GESTORA, CATEGORIA ECONÔMICA, ORIGEM RECEITA, ESPÉCIE RECEITA, DETALHAMENTO, VALOR PREVISTO ATUALIZADO, VALOR LANÇADO, VALOR REALIZADO, PERCENTUAL REALIZADO, DATA LANÇAMENTO, ANO EXERCÍCIO]
Index: []

Obtenemos un reporte de los valores nulos

In [46]:
reporte_brasil = sl.reporte(datos_brasil)
reporte_brasil

index  valores_nulos  porcentaje_nulos       tipo_dato
0       CÓDIGO ÓRGÃO SUPERIOR              0              0.00          object
1         NOME ÓRGÃO SUPERIOR              0              0.00          object
2                CÓDIGO ÓRGÃO              0              0.00          object
3                  NOME ÓRGÃO              0              0.00          object
4      CÓDIGO UNIDADE GESTORA              0              0.00          object
5        NOME UNIDADE GESTORA              0              0.00          object
6         CATEGORIA ECONÔMICA              0              0.00          object
7              ORIGEM RECEITA              0              0.00          object
8             ESPÉCIE RECEITA              0              0.00          object
9                DETALHAMENTO              0              0.00          object
10  VALOR PREVISTO ATUALIZADO              0              0.00         float64
11              VALOR LANÇADO              0              0.00         float64
12            VALOR REALIZADO              0              0.00         float64
13       PERCENTUAL REALIZADO              0              0.00         float64
14            DATA LANÇAMENTO            578              0.06  datetime64[ns]
15              ANO EXERCÍCIO              0              0.00           int64

La cantidad de valores nulos (todos en la columna "DATA LANÇAMENTO") es mínima, un 0.06 %. Para decidir si los podemos eliminar, vamos a analizar los valores de ingresos.

In [47]:
# creamos un dataframe filtrando los valores nulos
valores_nulos = datos_brasil[datos_brasil["DATA LANÇAMENTO"].isnull()]
# usamos .describe para obtener el resumen estadístico
valores_nulos.describe().T

count               mean     min         25%  \
VALOR PREVISTO ATUALIZADO  578.0                0.0     0.0         0.0   
VALOR LANÇADO              578.0  4495813047.449048     6.0  572100.915   
VALOR REALIZADO            578.0                0.0     0.0         0.0   
PERCENTUAL REALIZADO       578.0                0.0     0.0         0.0   
DATA LANÇAMENTO                0                NaT     NaT         NaT   
ANO EXERCÍCIO              578.0        2018.652249  2016.0      2017.0   

                                     50%             75%                  max  \
VALOR PREVISTO ATUALIZADO            0.0             0.0                  0.0   
VALOR LANÇADO              112386714.365  1499607661.085  257513407925.600006   
VALOR REALIZADO                      0.0             0.0                  0.0   
PERCENTUAL REALIZADO                 0.0             0.0                  0.0   
DATA LANÇAMENTO                      NaT             NaT                  NaT   
ANO EXERCÍCIO                     2019.0          2020.0               2021.0   

                                        std  
VALOR PREVISTO ATUALIZADO               0.0  
VALOR LANÇADO              16797766189.7325  
VALOR REALIZADO                         0.0  
PERCENTUAL REALIZADO                    0.0  
DATA LANÇAMENTO                         NaN  
ANO EXERCÍCIO                      1.719045

Nos interesa el VALOR REALIZADO, que es el monto realmente recaudado. Como en este resumen es todo cero, podemos suponer que, en las fechas nulas, no se recaudó el valor correspondiente.

Vamos a calcular qué porcentaje del total supondría el VALOR LANÇADO (importe registrado) 

In [48]:
porcentaje = round((((valores_nulos["VALOR LANÇADO"].sum())/(datos_brasil["VALOR LANÇADO"].sum()))*100), 2)
porcentaje

np.float64(48.29)

La suma de VALOR LANÇADO en las fechas nulas supondría un 48.29 % del total, por lo que los valores nulos se van a mantener.

- Detección y Corrección de Errores en Categorizaciones

In [49]:
datos_brasil.describe(include="O").T

count unique  \
CÓDIGO ÓRGÃO SUPERIOR   1026299     25   
NOME ÓRGÃO SUPERIOR     1026299     25   
CÓDIGO ÓRGÃO            1026299    291   
NOME ÓRGÃO              1026299    287   
CÓDIGO UNIDADE GESTORA  1026299    364   
NOME UNIDADE GESTORA    1026299    356   
CATEGORIA ECONÔMICA     1026299      5   
ORIGEM RECEITA          1026299     15   
ESPÉCIE RECEITA         1026299     63   
DETALHAMENTO            1026299   1886   

                                                                      top  \
CÓDIGO ÓRGÃO SUPERIOR                                               26000   
NOME ÓRGÃO SUPERIOR                                Ministério Da Educação   
CÓDIGO ÓRGÃO                                                        25000   
NOME ÓRGÃO              Ministério Da Economia - Unidades Com Vínculo ...   
CÓDIGO UNIDADE GESTORA                                             170013   
NOME UNIDADE GESTORA              Setorial Orcamentaria E Financeira / Me   
CATEGORIA ECONÔMICA                                    Receitas Correntes   
ORIGEM RECEITA                                  Outras Receitas Correntes   
ESPÉCIE RECEITA              Serviços Administrativos e Comerciais Gerais   
DETALHAMENTO                  SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

                          freq  
CÓDIGO ÓRGÃO SUPERIOR   361891  
NOME ÓRGÃO SUPERIOR     361891  
CÓDIGO ÓRGÃO            119811  
NOME ÓRGÃO              119811  
CÓDIGO UNIDADE GESTORA  108481  
NOME UNIDADE GESTORA    108481  
CATEGORIA ECONÔMICA     979596  
ORIGEM RECEITA          322907  
ESPÉCIE RECEITA         278157  
DETALHAMENTO            159039

Se puede ver que hay discrepancias entre el número de valores únicos de CÓDIGO ÓRGÃO (291) y  NOME ÓRGÃO (287) y CÓDIGO UNIDADE GESTORA (364) NOME UNIDADE GESTORA (356). Como en ambos casos hay más códigos que nombres y no hay valores nulos, esto puede deberse a que hay nombres que tienen más de un código asignado. Vamos a comprobar si hay nombres con más de un código

In [ ]:
# Agrupamos por NOME ÓRGÃO y contamos los códigos únicos usando .nunique(), que nos devuelve el número de valores únicos.
conteo_codigos_organo = datos_brasil.groupby('NOME ÓRGÃO')['CÓDIGO ÓRGÃO'].nunique().reset_index() 
# Filtramos para que solo muestre las filas que tengan un número de valores únicos >1
conteo_codigos_organo[conteo_codigos_organo['CÓDIGO ÓRGÃO'] > 1]

NOME ÓRGÃO  CÓDIGO ÓRGÃO
7      Agência Nacional De Transportes Aquaviários             2
121                      Fundo Nacional Antidrogas             2
131                        Fundo Nacional Do Idoso             2
133  Fundo Nacional Para A Criança E O Adolescente             2

Podemos ver que hay 4 órganos que tienen 2 códigos cada uno

In [ ]:
# Ahora agrupamos por NOME UNIDADE GESTORA y contamos los códigos únicos usando .nunique(), que nos devuelve el número de valores únicos.
conteo_codigos_unidad = datos_brasil.groupby('NOME UNIDADE GESTORA')['CÓDIGO UNIDADE GESTORA'].nunique().reset_index() 
# Filtramos para que solo muestre las filas que tengan un número de valores únicos >1
conteo_codigos_unidad[conteo_codigos_unidad['CÓDIGO UNIDADE GESTORA'] > 1]

NOME UNIDADE GESTORA  CÓDIGO UNIDADE GESTORA
8      Agencia Nacional De Transportes Aquaviarios                       2
60       Coordenacao-Geral De Orcamento E Financas                       2
63        Coordenacao-Geral De Recursos Logisticos                       2
79    Diretoria De Financas-Sistema Para O Pais/Mm                       2
153                      Fundo Nacional Antidrogas                       2
162                        Fundo Nacional Do Idoso                       2
164  Fundo Nacional Para A Crianca E O Adolescente                       2
292    Setorial Prog.Orcamentaria E Financeira-Sfb                       2

En este caso hay más nombres con 2 códigos.

Vamos a filtrar por uno de los órganos que tienen más de un código asignado para ver si puede ser un error.

In [75]:
datos_brasil[datos_brasil['NOME ÓRGÃO'].isin(["Fundo Nacional Do Idoso"])]

CÓDIGO ÓRGÃO SUPERIOR                            NOME ÓRGÃO SUPERIOR  \
3784                   57000  Ministério Das Mulheres, Igualdade Racial, Da   
3789                   57000  Ministério Das Mulheres, Igualdade Racial, Da   
3794                   57000  Ministério Das Mulheres, Igualdade Racial, Da   
3811                   57000  Ministério Das Mulheres, Igualdade Racial, Da   
3815                   57000  Ministério Das Mulheres, Igualdade Racial, Da   
...                      ...                                            ...   
109560                 81000  Ministério Da Mulher, Família E Direitos Huma   
109561                 81000  Ministério Da Mulher, Família E Direitos Huma   
109562                 81000  Ministério Da Mulher, Família E Direitos Huma   
109563                 81000  Ministério Da Mulher, Família E Direitos Huma   
109564                 81000  Ministério Da Mulher, Família E Direitos Huma   

       CÓDIGO ÓRGÃO               NOME ÓRGÃO CÓDIGO UNIDADE GESTORA  \
3784          64902  Fundo Nacional Do Idoso                 207001   
3789          64902  Fundo Nacional Do Idoso                 207001   
3794          64902  Fundo Nacional Do Idoso                 207001   
3811          64902  Fundo Nacional Do Idoso                 207001   
3815          64902  Fundo Nacional Do Idoso                 207001   
...             ...                      ...                    ...   
109560        30914  Fundo Nacional Do Idoso                 307002   
109561        30914  Fundo Nacional Do Idoso                 307002   
109562        30914  Fundo Nacional Do Idoso                 307002   
109563        30914  Fundo Nacional Do Idoso                 307002   
109564        30914  Fundo Nacional Do Idoso                 307002   

           NOME UNIDADE GESTORA CATEGORIA ECONÔMICA            ORIGEM RECEITA  \
3784    Fundo Nacional Do Idoso  Receitas Correntes  Transferências Correntes   
3789    Fundo Nacional Do Idoso  Receitas Correntes  Transferências Correntes   
3794    Fundo Nacional Do Idoso  Receitas Correntes       Receita Patrimonial   
3811    Fundo Nacional Do Idoso  Receitas Correntes  Transferências Correntes   
3815    Fundo Nacional Do Idoso  Receitas Correntes  Transferências Correntes   
...                         ...                 ...                       ...   
109560  Fundo Nacional Do Idoso  Receitas Correntes       Receita Patrimonial   
109561  Fundo Nacional Do Idoso  Receitas Correntes       Receita Patrimonial   
109562  Fundo Nacional Do Idoso  Receitas Correntes  Transferências Correntes   
109563  Fundo Nacional Do Idoso  Receitas Correntes  Transferências Correntes   
109564  Fundo Nacional Do Idoso  Receitas Correntes  Transferências Correntes   

                                      ESPÉCIE RECEITA  \
3784    Transferências dos Municípios e de suas Entid   
3789    Transferências de Outras Instituições Pública   
3794                              Valores Mobiliários   
3811    Transferências de Outras Instituições Pública   
3815    Transferências dos Municípios e de suas Entid   
...                                               ...   
109560                            Valores Mobiliários   
109561                            Valores Mobiliários   
109562              Transferências de Pessoas Físicas   
109563              Transferências de Pessoas Físicas   
109564        Transferências de Instituições Privadas   

                                         DETALHAMENTO  \
3784    Transferências dos Municípios e de suas Entid   
3789    Transferências de Outras Instituições Pública   
3794               REMUNERACAO DE DEPOSITOS BANCARIOS   
3811    Transferências de Outras Instituições Pública   
3815    Transferências dos Municípios e de suas Entid   
...                                               ...   
109560   REMUNERACAO DE DEPOSITOS BANCARIOS-PRINCIPAL   
109561   REMUNERACAO DE DEPOSITOS BANCARIOS-PRINCIPAL   
109562    TRANSFERENCIAS DE PESSOA

Hemos filtrado por "Fundo Nacional Do Idoso" y en el resultado podemos ver que tiene dos códigos, pero parece deberse a que depende de dos órganos superiores o que cada órgano superior tiene un órgano llamado así, se da el mismo caso con la unidad gestora:  
    1. 81000 - Ministério Da Mulher, Família E Direitos Huma... -- 30914 Fundo Nacional Do Idoso -- 307002 Fundo Nacional Do Idoso  
    2. 57000 - Ministério Das Mulheres, Igualdade Racial, Da	... -- 64902 Fundo Nacional Do Idoso -- 207001 Fundo Nacional Do Idoso  

Dado que no conocemos las interdependencias entre órganos superiores, órganos y unidades gestoras, no tenemos forma de saber si estos códigos son correctos o no, por lo que, en caso de ser necesario, se usará el nombre de la entidad correspondiente.

Comprobamos el formato de las categorías econónicas. Extraemos los valores únicos y, como solo hay 5, podemos comprobar visualmente si hay errores.

In [63]:
datos_brasil["CATEGORIA ECONÔMICA"].unique()

array(['Receitas Correntes', 'Receitas de Capital',
       'Receitas Correntes - intra-orçamentárias', 'Sem informação',
       'Receitas de Capital - intra-orçamentárias'], dtype=object)

Parece que no hay inconsistencias en la categoría económica.

## **Fase 3: Análisis Exploratorio de Datos (EDA)**

1. **Distribución de Ingresos por Categoría Económica:**

   - Analizar las categorías de ingresos más significativas y su participación en los ingresos totales.

   - Calcular la diferencia promedio entre ingresos previstos y realizados por cada categoría.

In [92]:
# Para analizar las categorías de ingresos agrupamos por CATEGORIA ECONÔMICA y sumamos el VALOR REALIZADO, ya que es el ingreso real. 
# Añadimos .reset_index(name = "ingreso realizado") para darle nombre a la columna y ordenamos por valores de forma descencente (ascending = False)
ingresos_categoria = datos_brasil.groupby("CATEGORIA ECONÔMICA")["VALOR REALIZADO"].sum().sort_values(ascending=False).reset_index(name = "ingreso realizado")
ingresos_categoria

CATEGORIA ECONÔMICA  ingreso realizado
0                         Receitas Correntes       1.290934e+13
1                        Receitas de Capital       1.240714e+13
2   Receitas Correntes - intra-orçamentárias       2.926518e+11
3  Receitas de Capital - intra-orçamentárias       2.825084e+10
4                             Sem informação       4.657000e+01

In [93]:
# Calculamos el porcentaje que supone cada categoría sobre los ingresos totales (suma de VALOR REALIZADO de todo el dataframe datos_brasil) y lo añadimos como columna nueva del dataframe ingresos_categoria
ingresos_categoria["porcentaje"] = round(((ingresos_categoria["ingreso realizado"]/datos_brasil["VALOR REALIZADO"].sum())*100),2)
ingresos_categoria

CATEGORIA ECONÔMICA  ingreso realizado  porcentaje
0                         Receitas Correntes       1.290934e+13       50.35
1                        Receitas de Capital       1.240714e+13       48.39
2   Receitas Correntes - intra-orçamentárias       2.926518e+11        1.14
3  Receitas de Capital - intra-orçamentárias       2.825084e+10        0.11
4                             Sem informação       4.657000e+01        0.00

Es evidente que las categorías "Receitas Correntes" y "Receitas de Capital" son las más significativas, ya que entre las dos suponen el 98.74 % de todos los ingresos, siendo "Receitas Correntes", con el 50.35% la que más participa, aunque la diferencia con "Receitas de Capital" es de apenas un 2%.

In [95]:
# Añadimos una nueva columna al df datos_brasil con el cálculo de la diferencia entre ingresos previstos y realizados por cada categoría.
# Usamos las columnas "VALOR PREVISTO ATUALIZADO" como ingreso previsto y "VALOR REALIZADO" para el ingreso previsto.
datos_brasil["diferencia_ingresos"] = datos_brasil["VALOR REALIZADO"] - datos_brasil["VALOR PREVISTO ATUALIZADO"]
# Añadimos una columna nueva a ingresos_categoría con la diferencia media por categoría
ingresos_categoria["diferencia_media"] = datos_brasil.groupby("CATEGORIA ECONÔMICA")["diferencia_ingresos"].mean()
ingresos_categoria

CATEGORIA ECONÔMICA  ingreso realizado  porcentaje  \
0                         Receitas Correntes       1.290934e+13       50.35   
1                        Receitas de Capital       1.240714e+13       48.39   
2   Receitas Correntes - intra-orçamentárias       2.926518e+11        1.14   
3  Receitas de Capital - intra-orçamentárias       2.825084e+10        0.11   
4                             Sem informação       4.657000e+01        0.00   

   diferencia_media  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN